doc/neps/nep-0055-string_dtype.rst

```diff
@@ -394,133 +399,271 @@ existing concept of an array of references in NumPy beyond the ``object`` DType
-will still have 7 bytes of space in the ``len`` field, which is much more than
-is likely to be necessary to store the length of a single array element in
-real-world use, and having 256 possibilities for flags gives us flexibility for
-the future.
+So it is only guaranteed to be the same size in memory as two ``size_t`` fields,
+with no guarantees about data layout or precisely how the data are
+stored. The raw string data can only be accessed once the packed string data are
+loaded into an unpacked string with the following layout:
+
+.. code-block:: C
+
+   struct npy_unpacked_static_string {
+       size_t size;
+       const char *buf;
+   };
 
-Besides the string data itself, each array element requires 16 bytes of storage
-for the ``npy_static_string`` instance in the array buffer. In principle we
-could use a 32 bit integer to store the ``len`` field, saving 4 bytes per array
-element, but if we only use a single bit for the small string optimization
-that will still leave us with an uncomfortably small 31 bits of space in the
-``len`` field. In addition, making use of the small string optimization will
-somewhat offset the memory cost of a 64 bit ``len`` field, since many real-world
-use-cases employ small strings.
-
-Missing Data
-++++++++++++
-
-By default, zeroed out entries in the array buffer represent empty
-strings. However, if the DType instance was created with an ``na_object`` field,
-zeroed-out entries represent missing data. By making this choice, a zero-filled
-newly allocated buffer returned by ``calloc`` does not need any additional
-post-processing to produce an empty array. This choice also means casts between
-different missing data representations are views.
-
-Whether or not the ``na_object`` is set, empty strings are not stored in the
-sidecar buffer since they require no additional storage besides the entry in the
-array buffer itself. This means that adjacent entries in the sidecar buffer are
-not necessarily adjacent entries in the array buffer.
+Where ``size`` is the size, in bytes, of the string and ``buf`` is a const
+pointer to the beginning of a UTF-8 encoded bytestream containing string
+data. This is a *read-only* view onto the string, we will not expose a public
+interface for modifying these strings. We do not append a trailing null
+character to the byte stream, so users attempting to pass the ``buf`` field to
+an API expecting a C string must create a copy with a trailing null. As a
+positive consequence, ``StringDType`` array entries can contain arbitrary
+embedded or trailing null characters.
+
+Internally, we represent the string as a union, with the following definition on
+little-endian architectures:
+
+.. code-block:: c
+
+   typedef struct _npy_static_vstring_t {
+      size_t offset;
+      size_t size_and_flags;
+   } _npy_static_string_t;
+
+   typedef struct _short_string_buffer {
+      char buf[sizeof(_npy_static_string_t) - 1];
+      unsigned char size_and_flags;
```

Comment:
Yeah, maybe... we haven't really used bit-fields much, but maybe that is just a remnant of a past where they didn't exist...

(In either case, I don't care much about it though, these are implementation details, that are OK to change.)

Reply:
> Might it make sense to define size and flags as separate bit fields?

Neat, I wasn't aware of this feature. I'm not sure it actually helps much besides making the code clearer in this case since we're packing the bits into a `char` anyway, so using a bitfield doesn't save us any RAM, it just makes it explicit what each of the bits are for.

> you do not seem to worry about saving this information

In principle we could use them for serialization but right now we don't save the bitflags.

Reply:
For now I'd like to focus on other things, would you be ok with coming back to this later as a refactor? I definitely agree clarity is important, I just have a lot of stuff that I'd like to get done in time for NumPy 2.0!